python/pysmurf/core/utilities/_SmurfPublisher.py

```diff
@@ -17,27 +17,179 @@
+        self.log_stop()
+
+    def _getoptenv(self, key, options, default=None):
+        """Look for `key` in dict `options`, and if it's not found then try to
+        get it from the environment (including the local `env_root`
+        prefix), and if it's not found then return the `default`.
+
+        """
+        if key in options:
+            return options[key]
+        return os.getenv(self.env_root + key, default)
+
+    # Backend send functions.  One of these will be assigned to
+    # self._backend.
+
+    def _backend_null(self, json_msg):
+        pass
+
+    def _backend_udp(self, json_msg):
+        payload = bytes(json_msg, 'utf_8')
+        if len(payload) > UDP_MAX_BYTES:
+            # Can't send this; write to stderr and notify consumer.
+            error = 'Backend error: dropped large UDP packet (%i bytes).' % len(payload)
+            sys.stderr.write('%s %s' % (self, error))
+            self.publish({'message': error}, 'backend_error')
+            return
+        self.udp_sock.sendto(payload, (self.udp_ip, self.udp_port))
+
+    # Publishing functions -- external interface.
+
+    def publish(self, data, msgtype='general'):
+        # Create the wrapper.
+        output = {
+            'host': self.host,
+            'id': self.pub_id,
+            'script': self.script_id,
+            'time': time.time(),
+            'seq_no': self.seq_no,
+            'type': msgtype,
+        }
+        self.seq_no += 1
+        # Add in the data and convert to json.
+        output['payload'] = data
+        jtext = json.dumps(output)
+        # Send.
+        self._backend(jtext)
+
+    def log(self, message):
+        return self.publish({'message': message}, 'log')
+
+    def log_start(self):
+        """Publishes start message"""
+        return self.publish({}, 'start')
+
+    def log_stop(self):
+        """Publishes stop message"""
+        return self.publish({}, 'stop')
+
+    def register_file(self, path, type, format='', timestamp=None, plot=False):
+        """
+        Publishes file info so it can be picked up by the pysmurf-archiver.
+
+        Args:
+        -----
+        path (str):
+            full path to file.
+        type (str):
+            Type of data file, e.g. "tuning" or "config_snapshot"
+        format (str):
+            File extension. E.g. "npy" or "txt"
+        timestamp (float):
+            Unix timestamp when file was created.
+        plot (bool):
+            True if file is a plot
+        """
+        if timestamp is None:
+            timestamp = time.time()
+
+        file_data = {
+            'path': path,
+            'type': type,
+            'format': format,
+            'timestamp': timestamp,
+            'plot': plot,
+            'pysmurf_version': 'replace_with_version'
+        }
+
+        return self.publish(file_data, 'data_file')
+
     def _varListen(self, path, varVal):
         self.pub.publish(data=f'{path}={varVal.value}',msgType='general')
```

Comment:
`self` doesn't have a `pub` method. Should this call `self.publish` instead? 
Also, the `publish` method doesn't have a `msgType` argument, but instead a `msgtype`. 

So, I think this line should be:
```
self.publish(data=f'{path}={varVal.value}',msgtype='general')
```

Reply:
Ok these have all been addressed.